File: carbondesign/tags/slider.py
"""
Slider
======

See: https://www.carbondesignsystem.com/components/slider/usage/

Sliders provide a visual indication of adjustable content, where the user can
increase or decrease the value by moving the handle along a horizontal track.

Overview
--------

The slider in its basic form should be accompanied by a label and a number
input that doubles as a display for the slider’s current value.

The basic slider does not include discrete values, as the slider represents
a percentage of 0-100. In this case it is not necessary for a user to choose
a specific value, but instead generally increase or decrease an input. For
example, the user increases the slider amount and the volume of the music gets
louder.

The more complex versions should be used for selecting a specific value within
a value range.
""" # pylint:disable=line-too-long
# pylint:disable=too-many-lines

from django.utils.translation import gettext as _
#-
from .base import FormNode

class Slider(FormNode):
    """Slider component.
    """
    NODE_PROPS = ('id', 'min', 'max', 'step', 'light')
    "Extended Template Tag arguments."
    CLASS_AND_PROPS = ('label', 'slider')
    "Prepare xxx_class and xxx_props values."

    def prepare(self, values, context):
        """Prepare values for rendering the templates.
        """
        values['txt_slider'] = _("slider")
        values['min'] = self.eval(self.kwargs.get('min', 0), context)
        values['max'] = self.eval(self.kwargs.get('max', 100), context)
        values['step'] = self.eval(self.kwargs.get('step', 1), context)
        values['value'] = self.bound_field.value()

        if self.eval(self.kwargs.get('light'), context):
            values['class'].append('bx--text-input--light')

        if self.eval(self.kwargs.get('disabled'), context):
            values['label_class'].append('bx--label--disabled')
            values['slider_class'].append('bx--slider--disabled')


    def render_default(self, values, context):
        """Output html of the component.
        """
        template = """
<div class="bx--form-item">
  {tmpl_label}
  <div class="bx--slider-container">
    <label id="bottom_range-label-{id}" class="bx--slider__range-label">
      {min}
    </label>
    <div class="bx--slider {slider_class}" data-slider
        data-slider-input-box="#{id}" {slider_props}>
      <div class="bx--slider__thumb" tabindex="0"></div>
      <div class="bx--slider__track"></div>
      <div class="bx--slider__filled-track"></div>
      <input aria-label="{txt_slider}" id="slider-{id}"
          class="bx--slider__input" type="range" step="{step}" min="{min}"
          max="{max}" value="{value}">
    </div>
    <label id="top_range-label-{id}" class="bx--slider__range-label">
      {max}
    </label>
    <input id="{id}"
        aria-labelledby="bottom_range-label-{id} top_range-label-{id}"
        type="number" class="bx--text-input bx--slider-text-input {class}"
        placeholder="{min}" value="{value}" {props}>
  </div>
</div>
"""
        return self.format(template, values, context)


components = {
    'Slider': Slider,
}

File: carbondesign/mithril-js/slider.js
import m from 'mithril/hyperscript';
//-
import { FormNode } from './base';

export class Slider extends FormNode
{
    NODE_PROPS = ['min', 'max', 'step', 'light']

    prepare(vnode, values, context)
    {
        values.txt_slider = gettext("slider");
        values.min = vnode.attrs.min || 0;
        values.max = vnode.attrs.max || 100;
        values.step = vnode.attrs.step || 1;
        values.value = this.bound_field.value();

        if (vnode.attrs.light)
        {
            values['class'].push('bx--slider-text-input--light');
        }

        if (vnode.attrs.disabled)
        {
            values.label_class.push('bx--label--disabled');
            values.wrapper_class.push('bx--slider--disabled');
        }
    }

    render_default(vnode, values, context)
    {
        return (
//##
m('div.bx--form-item', null,
  [
    m('label',
      {
        'class': `bx--label ${values.label_class}`,
        ...values.label_props,
      },
      values.label),
    m('div.bx--slider-container', null,
      [
        m('label.bx--slider__range-label',
          {
            id: `slider-input-box_bottom-range-label-${values.id}`,
          },
          values.min),
        m('div',
          {
            'class': `bx--slider ${values.wrapper_class}`,
            'data-slider': '',
            'data-slider-input-box': `#${values.id}`,
          },
          [
            m('div.bx--slider__thumb', {tabindex: 0}),
            m('div.bx--slider__track'),
            m('div.bx--slider__filled-track'),
            m('input.bx--slider__input',
              {
                'aria-label': values.txt_slider,
                id: `slider-${values.id}`,
                type: 'range',
                step: values.step,
                min: values.min,
                max: values.max,
                value: values.value,
          ]),
        m('label.bx--slider__range-label',
          {
            id: `slider-input-box_top-range-label-${values.id}`,
          },
          values.max),
        m('input',
          {
            id: values.id,
            'aria-labelledby': `slider-input-box_bottom-range-label-${values.id} slider-input-box_top-range-label-${values.id}`,
            type: 'number',
            'class': `bx--text-input bx--slider-text-input ${values['class']}`,
            placeholder: values.min,
            value: values.value,
          }),
      ]),
  ])
//##
        );
    }
}
